File: csharp/src/Training.Module8/BoyScout/After/ReleaseSummaryFormatter.cs
using System.Text;

namespace Training.Module8.BoyScout.After;

public sealed class ReleaseSummaryFormatter
{
    public string Format(
        string releaseId,
        IReadOnlyList<DeploymentResult> results)
    {
        Validate(releaseId, results);

        var summary = new StringBuilder()
            .Append("Release ")
            .Append(releaseId.Trim())
            .Append('\n');
        int successfulDeployments = 0;
        foreach (DeploymentResult result in results)
        {
            ArgumentNullException.ThrowIfNull(result, "result");
            summary.Append(FormatResult(result)).Append('\n');
            if (result.Status == DeploymentStatus.Success)
            {
                successfulDeployments++;
            }
        }
        return summary.Append("Successful: ")
            .Append(successfulDeployments)
            .Append('/')
            .Append(results.Count)
            .ToString();
    }

    private static void Validate(
        string releaseId,
        IReadOnlyList<DeploymentResult> results)
    {
        ArgumentNullException.ThrowIfNull(releaseId, "releaseId");
        ArgumentNullException.ThrowIfNull(results, "results");
        if (string.IsNullOrWhiteSpace(releaseId))
        {
            throw new ArgumentException(
                "releaseId must not be blank");
        }
        if (results.Count == 0)
        {
            throw new ArgumentException(
                "results must not be empty");
        }
    }

    private static string FormatResult(DeploymentResult result)
    {
        string label = result.Status == DeploymentStatus.Success
            ? "[OK]"
            : "[ERROR]";
        return label + " " + result.Environment + ": "
            + result.Description;
    }
}

File: typescript/src/module8/boyscout/after/ReleaseSummaryFormatter.ts
import { IllegalArgumentError } from '../../../shared/errors.js';
import { requireNonNull } from '../../../shared/requireNonNull.js';
import type { DeploymentResult } from '../DeploymentResult.js';
import { DeploymentStatus } from '../DeploymentStatus.js';

export class ReleaseSummaryFormatter {
  format(releaseId: string, results: readonly DeploymentResult[]): string {
    ReleaseSummaryFormatter.validate(releaseId, results);

    const summary: string[] = [`Release ${releaseId.trim()}\n`];
    let successfulDeployments = 0;
    for (const result of results) {
      requireNonNull(result, 'result');
      summary.push(ReleaseSummaryFormatter.formatResult(result), '\n');
      if (result.status === DeploymentStatus.SUCCESS) {
        successfulDeployments++;
      }
    }
    summary.push(`Successful: ${successfulDeployments}/${results.length}`);
    return summary.join('');
  }

  private static validate(releaseId: string, results: readonly DeploymentResult[]): void {
    requireNonNull(releaseId, 'releaseId');
    requireNonNull(results, 'results');
    if (releaseId.trim().length === 0) {
      throw new IllegalArgumentError('releaseId must not be blank');
    }
    if (results.length === 0) {
      throw new IllegalArgumentError('results must not be empty');
    }
  }

  private static formatResult(result: DeploymentResult): string {
    const label = result.status === DeploymentStatus.SUCCESS ? '[OK]' : '[ERROR]';
    return `${label} ${result.environment}: ${result.description}`;
  }
}
